extensions/amp-story/1.0/history.js

```diff
@@ -0,0 +1,157 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dict} from '../../../src/utils/object';
+import {getState} from '../../../src/history';
+import {parseJson} from '../../../src/json';
+
+const EXPIRATION_DURATION_MILLIS = 10 * 60 * 1000; // 10 Minutes
+const CREATION_TIME = 'time';
+const STATE = 'state';
+/** @visibleForTesting */
+export const LOCAL_STORAGE_KEY = 'amp-story-state';
+
+/** @enum {string} */
+export const HistoryState = {
+  ATTACHMENT_PAGE_ID: 'ampStoryAttachmentPageId',
+  BOOKEND_ACTIVE: 'ampStoryBookendActive',
+  NAVIGATION_PATH: 'ampStoryNavigationPath',
+};
+
+/**
+ * Updates the value for a given state in the window history.
+ * Also persists state in localStorage for 10 minutes, so that state is
+ * persisted if Stories are shown within viewers that may recycle
+ * iframes or webviews.
+ * @param {!Window} win
+ * @param {string} stateName
+ * @param {string|boolean|Array<string>|null} value
+ */
+export function setHistoryState(win, stateName, value) {
+  const {history} = win;
+  const state = getState(history) || {};
+  const newHistory = {
+    ...state,
+    [stateName]: value,
+  };
+
+  history.replaceState(newHistory, '');
+  setLocalStorageState(win, newHistory);
+}
+
+/**
+ * Returns the value of a given state of the window history or localStorage.
+ * See `setHistoryState` for an explanation of the use of localStorage.
+ * @param {!Window} win
+ * @param {string} stateName
+ * @return {string|boolean|Array<string>|null}
+ */
+export function getHistoryState(win, stateName) {
+  const {history} = win;
+  if (history) {
```

Comment:
If `history` is not supporter, can we still try to get the state from localStorage?

Reply:
Maybe we can instead delete the if check. Which browser would not have this?